# Salmonella-enteric

Foodborne illness caused by a variety of contaminated foods and beverages continue to pose a major global public health problem. In the US, 48 million (about 1 in 6) people get infected due to foodborne illness, resulting in 128,000 hospitalizations and 3,000 deaths each year. GenomeTrakr Whole Genome Sequencing (WGS) Network, provided by the FDA's Center for Food Safety and Applied Nutrition (CFSAN), is then used as a common tool concerning the food safety. Salmonella enterica is one of the pathogens causing foodborne illness with the most infected individuals. Many studies have investigated the foodborne outbreak caused by Salmonella enteritidis using WGS approach. 

In this project, we will primarily concentrate on Salmonella enterica serovar detection to see if it can give us some information to identify the outbreak and potential transmission trend. 

## Data Source and Collection

We are using the Salmonella entreica data from [National Center for Biotechnology Informationâ€™s(NCBI) Pathogens Detection database](https://www.ncbi.nlm.nih.gov/pathogens/) in our study to explore the high-risk isolates of salmonella and predict the probability of different serovars of salmonella most likely to cause an outbreak at specific times. 

The NCBI Pathogens detection system contains comprehensive information on bacterial and fungal pathogen genomic sequences. Numerous sequenced samples from clinical cases and environmental and food sources are submitted by public health institutions and researchers worldwide to NCBI for active, real-time monitoring of pathogens and potential foodborne illness outbreaks. NCBI analyzes and compares these sequences to those in the NCBI Pathogens detection database. By identifying the related sequences, the NCBI can detect closely related isolates and thus investigate their concerning potential pathogen-related outbreaks.

The [system](https://www.ncbi.nlm.nih.gov/pathogens/about/) helps public health researchers to identify transmission chains in potential foodborne illness outbreaks by clustering related pathogen genome sequences and allows them to track resistance genes by providing information on antimicrobial resistance, stress response, and virulence genes found in bacterial genomic sequences using AMRFinderPlus. 

For complete dataset, [see](https://drive.google.com/file/d/1kmlia5vTlS_2r9Hp5TpCGLK7Uq3U02_p/view?usp=sharing)
